src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/ReconstructWork.tpp

```diff
@@ -0,0 +1,302 @@
+      make_not_null(
+          &get<ValenciaDivClean::Tags::TildeB<Frame::Inertial>>(*vars_on_face)),
+      make_not_null(&get<ValenciaDivClean::Tags::TildePhi>(*vars_on_face)),
+      rest_mass_density, specific_internal_energy, specific_enthalpy, pressure,
+      spatial_velocity, lorentz_factor,
+      get<hydro::Tags::MagneticField<DataVector, 3, Frame::Inertial>>(
+          *vars_on_face),
+      get<gr::Tags::SqrtDetSpatialMetric<DataVector>>(*vars_on_face),
+      get<gr::Tags::SpatialMetric<3>>(*vars_on_face),
+      get<hydro::Tags::DivergenceCleaningField<DataVector>>(*vars_on_face));
+}
+
+template <typename PrimsTags, typename TagsList, size_t ThermodynamicDim,
+          typename F>
+void reconstruct_prims_work(
+    const gsl::not_null<std::array<Variables<TagsList>, 3>*> vars_on_lower_face,
+    const gsl::not_null<std::array<Variables<TagsList>, 3>*> vars_on_upper_face,
+    const F& reconstruct, const Variables<PrimsTags>& volume_prims,
+    const EquationsOfState::EquationOfState<true, ThermodynamicDim>& eos,
+    const Element<3>& element,
+    const FixedHashMap<maximum_number_of_neighbors(3) + 1,
+                       std::pair<Direction<3>, ElementId<3>>,
+                       evolution::dg::subcell::NeighborData,
+                       boost::hash<std::pair<Direction<3>, ElementId<3>>>>
+        neighbor_data,
+    const Mesh<3>& subcell_mesh, size_t ghost_zone_size) noexcept {
+  using prim_tags_for_reconstruction =
+      tmpl::list<hydro::Tags::RestMassDensity<DataVector>,
+                 hydro::Tags::Pressure<DataVector>,
+                 hydro::Tags::LorentzFactorTimesSpatialVelocity<DataVector, 3>,
+                 hydro::Tags::MagneticField<DataVector, 3>,
+                 hydro::Tags::DivergenceCleaningField<DataVector>>;
+
+  ASSERT(Mesh<3>(subcell_mesh.extents(0), subcell_mesh.basis(0),
+                 subcell_mesh.quadrature(0)) == subcell_mesh,
+         "The subcell mesh should be isotropic but got " << subcell_mesh);
+  const size_t volume_num_pts = subcell_mesh.number_of_grid_points();
+  const size_t reconstructed_num_pts =
+      (subcell_mesh.extents(0) + 1) *
+      subcell_mesh.extents().slice_away(0).product();
+  const size_t nhbr_num_pts =
+      ghost_zone_size * subcell_mesh.extents().slice_away(0).product();
+  size_t vars_in_nhbr_count = 0;
+  tmpl::for_each<prim_tags_for_reconstruction>(
+      [&element, &neighbor_data, nhbr_num_pts, &reconstruct,
+       reconstructed_num_pts, volume_num_pts, &volume_prims,
+       &vars_in_nhbr_count, &vars_on_lower_face, &vars_on_upper_face,
+       &subcell_mesh](auto tag_v) noexcept {
+        using tag = tmpl::type_from<decltype(tag_v)>;
+        const typename tag::type* volume_tensor_ptr = nullptr;
+        Variables<tmpl::list<
+            hydro::Tags::LorentzFactorTimesSpatialVelocity<DataVector, 3>>>
+            lorentz_factor_times_v_I{};
+        if constexpr (std::is_same_v<
```

Comment:
Can you add a comment explaining why you have this `if tag == Wv^i` logic? Is it because this "product" tag isn't a compute item and so you have to set it by hand?